src/mods/storytimer.cpp

```diff
@@ -0,0 +1,427 @@
+#include "storytimer.h"
+
+#include "mkb/mkb.h"
+
+#include "mods/freecam.h"
+#include "systems/assembly.h"
+#include "systems/pad.h"
+#include "systems/pref.h"
+#include "utils/draw.h"
+#include "utils/patch.h"
+#include "utils/timerdisp.h"
+
+namespace storytimer {
+
+    enum class FullgameTimerOptions {
+        F_DontShow = 0,
+        F_AlwaysShow = 1,
+        F_BetweenWorlds = 2,
+        F_EndOfRun = 3,
+    };
+    
+    enum class SegmentTimerOptions {
+        S_DontShow = 0,
+        S_AlwaysShow = 1,
+        S_BetweenWorlds = 2,
+        S_EndOfRun = 3,
+    };
+
+static u32 s_spin_in_timer;
+static u32 s_gameplay_timer;
+static u32 s_postgoal_timer;
+static u32 s_stage_select_timer;
+static u32 s_exit_game_timer;
+static u32 s_fallout_timer;
+static u32 s_timeover_timer;
+static u32 s_spin_in_timer_correction;
+static u32 s_game_scenario_return_timer_correction;
+static u32 s_world_start_timer_correction;
+static u32 s_loadless_story_timer;
+static bool s_in_story;
+static bool s_is_on_world[11]; 
+static bool s_is_between_worlds;
+static bool s_is_run_complete;
+static bool s_is_on_spin_in;
+static bool s_is_on_gameplay;
+static bool s_is_postgoal;
+static bool s_is_on_stage_select_screen; 
+static bool s_is_on_exit_game_screen;
+static bool s_is_on_fallout_screen;
+static bool s_is_timeover;
+static bool s_can_increment_stage_counter;
+static bool s_lower_stage_counter;
+static bool s_start_stage_fade_out_timer;
+static u32 s_stage_fade_out_timer;
+static u32 s_prev_completed_stage_count;
+static s32 s_completed_stages;
+static u32 s_segment_timer[11]; // IW timer for world k
+static u32 s_split[11];    // s_split[k] is the loadless time on tape break of the 10th stage of world k
+static u32 s_segment_start_time[11];  // the loadless time at the start of world k, used to calculate s_segment_timer[k]
+static bool s_can_change_segment_start_time[11];
+static bool s_display_story_timer;
+static bool s_display_segment_timer;
+static u32 s_fullgame_timer_location_y;
+static u32 s_segment_timer_location_y;
+static constexpr s32 fullgame_timer_location_x = 18+24;
+static constexpr s32 fullgame_timer_text_offset = 56;
+static constexpr s32 segment_timer_location_x = 30+24;
+static constexpr s32 segment_timer_text_offset = 44;
+static constexpr s32 IW_time_location_x = 42+24; 
+static constexpr s32 IW_time_text_offset = 32;
+static constexpr s32 stage_fade_out_time = 49;
+    
+void tick() {
+    
+    // for later use, it's useful to record how many stages we've completed
+    // there is likely a simpler way to increment the stage counter in a way that works when you retry after breaking the tape and also does not break if you pause during 
+    // game goal init or game goal main
```

Comment:
I chose to rewrite the existing code without using 'mkb::get_world_unbeaten_stage_count' because that increments after the fade to white happens and not on tape break. My code for completed stages does increment on tape break though.  I'm not sure if my rewrite is simpler, or if using 'mkb::get_world_unbeaten_stage_count' and when I need something involving the number of completed stages to happen on tape break to just check if we are in GAME_GOAL_INIT and if we are not paused. I can try out the 2nd option if you think it would simplify the code further